Sources/Subjects/ReplaySubject.swift

```diff
@@ -37,53 +37,74 @@ public final class ReplaySubject<Output, Failure: Error>: Subject {
     }
 
     public func send(_ value: Output) {
-        lock.lock()
-        defer { lock.unlock() }
+        let subscriptions: [Subscription<AnySubscriber<Output, Failure>>]
 
-        guard isActive else { return }
+        do {
+          lock.lock()
+          defer { lock.unlock() }
 
-        buffer.append(value)
+          guard isActive else { return }
 
-        if buffer.count > bufferSize {
+          buffer.append(value)
+          if buffer.count > bufferSize {
             buffer.removeFirst()
+          }
+
+          subscriptions = self.subscriptions
         }
 
         subscriptions.forEach { $0.forwardValueToBuffer(value) }
     }
 
     public func send(completion: Subscribers.Completion<Failure>) {
-        lock.lock()
-        defer { lock.unlock() }
+        let subscriptions: [Subscription<AnySubscriber<Output, Failure>>]
+
+        do {
+            lock.lock()
+            defer { lock.unlock() }
 
-        guard isActive else { return }
+            guard isActive else { return }
 
-        self.completion = completion
+            self.completion = completion
+
+            subscriptions = self.subscriptions
+        }
 
         subscriptions.forEach { $0.forwardCompletionToBuffer(completion) }
     }
 
     public func send(subscription: Combine.Subscription) {
-        lock.lock()
-        defer { lock.unlock() }
-
         subscription.request(.unlimited)
     }
 
     public func receive<Subscriber: Combine.Subscriber>(subscriber: Subscriber) where Failure == Subscriber.Failure, Output == Subscriber.Input {
-        lock.lock()
-        defer { lock.unlock() }
-        
         let subscriberIdentifier = subscriber.combineIdentifier
 
-        let subscription = Subscription(downstream: AnySubscriber(subscriber)) { [weak self] in
-            guard let self = self,
-                  let subscriptionIndex = self.subscriptions
-                                              .firstIndex(where: { $0.innerSubscriberIdentifier == subscriberIdentifier }) else { return }
+        let subscription: Subscription<AnySubscriber<Output, Failure>>
+        let buffer: [Output]
+        let completion: Subscribers.Completion<Failure>?
 
-            self.subscriptions.remove(at: subscriptionIndex)
-        }
+        do {
+            lock.lock()
+            defer { lock.unlock() }
+
+            subscription = Subscription(downstream: AnySubscriber(subscriber)) { [weak self] in
+                guard let self = self else { return }
```

Comment:
I wonder if the body of this completion handler could be extracted to a separate method? I think having it nested here makes it a little difficult to reason about the two nested scopes where `lock` is referenced. They appear to overlap, but are actually separated in time.

Additionally, initializing the local `subscription` variable doesn't need to be protected by the lock, only the part where it's appended to `self.subscriptions`.

Reply:
> Additionally, initializing the local `subscription` variable doesn’t need to be protected by the lock, only the part where it’s appended to `self.subscriptions`.

Great catch. Fixing in the next commit. ✅

> I wonder if the body of this completion handler could be extracted to a separate method?

My only hesitation with refactoring as a method would be the implicit `self` capture — moving `subscription`’s initialization out of the `do` block should hopefully make the locking across time more clear.

Reply:
I wasn't thinking that capturing `self` would be implicit — something more along these lines:

```swift
let subscription = Subscription(downstream: AnySubscriber(subscriber)) { [weak self] in
  self?.completeSubscriber(withIdentifier: subscriberIdentifier)
}
```

Reply:
Ahh roger roger, I dig it! Will update in a few. 👌🏽